src/vuln_analysis/utils/chain_of_calls_retriever.py

```diff
@@ -271,6 +338,195 @@ def get_possible_docs(self, function_name_to_search: str, package: str, exclusio
+        while q:
+            target_doc = q.popleft()
+
+            target_pkg = self.language_parser.get_package_names(target_doc)[0]
+            if target_doc.metadata.get('state') == "invalid":
+                logger.debug("get_relevant_documents: invalid function %s", target_doc.metadata['source'])
+                continue
+            function_name = self.language_parser.get_function_name(target_doc)
+        
+            function_file = target_doc.metadata.get('source')
+            hashed_value = calculate_hashable_string_for_function(function_file, function_name)
+            if hashed_value in self.last_visited:
+                continue
+
+            self.last_visited[hashed_value] = 1
+            logger.debug("%d:file:%s, func_name : %s , pkg:%s queue len %d", 
+                        file_counter, target_doc.metadata['source'], function_name, target_pkg, len(q))
+
+            file_counter += 1
+            # Find a caller function and containing package in the dependency tree according to hierarchy
+            sub_start = time.time()
+            found_documents = self.__find_caller_functions(document_function=target_doc,
+                                                         function_package=target_pkg)
+            sub_elapsed = time.time() - sub_start
+            print(f"[PROFILE] __find_caller_functions took {sub_elapsed:.3f} seconds")
+            # If found, then add it to path
+            if found_documents is not None:
+                for doc_candidate in found_documents:                                        
+                    # If the function is in the application ( root package), then we finished and found such a path.
+                    if self.language_parser.is_root_package(doc_candidate):
+                        matching_documents.append(doc_candidate)
+                        self.found_path = True
+                        break
+                    # Otherwise, we continue to search for callers for the current found function, in order to extend
+                    # the chain of calls and potentially find a path from application to the vulnerable
+                    # function in input package
+                    else:
+                        q.append(doc_candidate)
+            
+            if self.found_path:
+                break
+        loop_elapsed = time.time() - loop_start
+        print(f"[PROFILE] Main loop in get_relevant_C_documents took {loop_elapsed:.3f} seconds")
+        # When the loop is finished, return list of documents ( path) and boolean indicating whether a path was
+        # found or not.        
+        logger.debug("get_relevant_documents: result %s", self.found_path)
+        logger.debug("get_relevant_documents: docs %s", matching_documents)
+        return matching_documents, self.found_path
+        
+    def depth_first_search(self, matching_documents, target_function_doc, 
+                          current_package_name) -> tuple[List[Document], bool]:
```

Comment:
@RedTanny Please make it private as it doesn't make sense to use it outside of here
```suggestion
    def __depth_first_search(self, matching_documents, target_function_doc, 
                          current_package_name) -> tuple[List[Document], bool]:
```